### ISSUE TITLE
Be specific about AMI update procedures

### ISSUE THREAD

**Issue by b-long** [Bioconductor/coursehelper#2]
In [this section](https://github.com/Bioconductor/coursehelper/tree/1c3bbf0b5a7d18406a0b2927113a2fcc8e83b461#building-a-course-ami) of the (administrator) documentation, we say : 
> _Update all R packages on the AMI by calling biocLite() without arguments. If R itself is outdated you might update it as well._

The process of doing this (as root?  as ubuntu?)  should be captured.  I'd suggest here: https://github.com/Bioconductor/Bioconductor/blob/master/documentation/AMI-management.md

**Comment by b-long** [comment]
In addition to the general need for documentation.  I found that on the current 3.2 AMI ([ami-85d88de0](https://console.aws.amazon.com/ec2/home?region=us-east-1#launchAmi=ami-85d88de0)), the following problem exists.  

R at the command line, and R via the R studio web interface are both run as the `ubuntu` user.  This user has 2 libraries : 
```
> .libPaths()
[1] "/home/ubuntu/R-libs"      "/usr/local/lib/R/library"
```
Unfortunately, the user-space directory contains content with varying permissions  : 
```
ubuntu@ip-172-30-0-145:~/R-libs$ ls -l
total 1424
drwxrwxr-x  6 ubuntu root   4096 May 28  2015 abind
drwxrwxr-x  7 ubuntu root   4096 May 28  2015 acepack
drwxrwxr-x 10 root   root   4096 Oct 14 18:47 affxparser
drwxrwxr-x 11 root   root   4096 Oct 14 18:55 affy
drwxrwxr-x  7 root   root   4096 Oct 14 18:51 affyio
drwxrwxr-x  8 root   root   4096 Oct 14 19:10 affylmGUI
drwxrwxr-x  8 root   root   4096 Oct 14 19:08 affyPLM
drwxrwxr-x 12 ubuntu root   4096 May 28  2015 amap
drwxrwxr-x  8 root   root   4096 Oct 14 19:05 annaffy
drwxrwxr-x 10 root   root   4096 Oct 14 18:59 annotate
drwxrwxr-x 11 root   root   4096 Oct 14 18:58 AnnotationDbi
drwxrwxr-x 13 root   root   4096 Oct 14 19:05 AnnotationForge
drwxrwxr-x  9 root   root   4096 Oct 14 18:59 AnnotationHub
drwxrwxr-x 11 ubuntu root   4096 May 28  2015 aroma.affymetrix
drwxrwxr-x  6 ubuntu root   4096 May 28  2015 aroma.apd
drwxrwxr-x 10 ubuntu root   4096 May 28  2015 aroma.core
drwxrwxr-x  8 root   root   4096 Oct 14 18:47 aroma.light
drwxrwxr-x  7 ubuntu root   4096 May 28  2015 assertthat
drwxrwxr-x  7 root   root   4096 Oct 14 18:47 base64enc
drwxrwxr-x  6 ubuntu root   4096 May 28  2015 BatchExperiments
drwxrwxr-x  8 ubuntu root   4096 May 28  2015 BatchJobs
drwxrwxr-x 10 root   root   4096 Oct 14 18:58 BayesPeak
drwxrwxr-x  8 root   root   4096 Oct 14 19:06 baySeq
drwxrwxr-x  7 ubuntu root   4096 May 28  2015 BBmisc
drwxrwxr-x  6 ubuntu root   4096 May 28  2015 BH
drwxrwxr-x  9 ubuntu root   4096 May 28  2015 BiasedUrn
drwxrwxr-x  7 ubuntu root   4096 May 28  2015 biglm
drwxrwxr-x 14 root   root   4096 Oct 14 18:51 Biobase
drwxrwxr-x  7 root   root   4096 Oct 14 18:47 BiocGenerics
drwxrwxr-x  8 root   root   4096 Oct 14 18:16 BiocInstaller
drwxrwxr-x  9 root   root   4096 Oct 14 18:47 BiocParallel
drwxrwxr-x 10 root   root   4096 Oct 14 18:47 BiocStyle
drwxrwxr-x  8 root   root   4096 Oct 14 18:59 biomaRt
drwxrwxr-x 12 root   root   4096 Oct 14 18:59 Biostrings
drwxrwxr-x 11 root   root   4096 Oct 14 19:17 biovizBase
drwxrwxr-x  8 ubuntu root   4096 May 28  2015 bit
drwxrwxr-x  7 ubuntu root   4096 May 28  2015 bitops
drwxrwxr-x  8 ubuntu ubuntu 4096 Jul  7  2015 boot

### Remainder omitted for brevity
```
As this seemed an obvious issue (it was causing failures when updating packages) - I decided to `chown` (recursively) everything in `/home/ubuntu/R-libs/`:
```
ubuntu@ip-172-30-0-145:~/R-libs$ sudo chown -R ubuntu:ubuntu .
```
This allowed me to open an R session (at the command line, as Ubuntu) and successfully update all packages using `biocLite()`.  It seemed everything was working properly.  Because I had confidence in this, I created a new 3.2 AMI based on my now modified instance.  Here's the updated version: [ami-c2e7c2a8](https://console.aws.amazon.com/ec2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-c2e7c2a8;sort=desc:creationDate)

Unfortunately, it seems I created the new AMI too quickly.  After creating the image, I did further testing of it via the Rstudio web interface.  This time, there's a prompt for a different user-space library :
```
Bioconductor version 3.2 (BiocInstaller 1.20.0), ?biocLite for help
> biocLite()
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.2 (BiocInstaller 1.20.1), R 3.2.2 (2015-08-14).
Old packages: 'checkmate', 'digest', 'MASS', 'Matrix', 'mgcv', 'stringi'
Update all/some/none? [a/s/n]: 
a
Warning in install.packages(update[instlib == l, "Package"], l, contriburl = contriburl,  :
  'lib = "/usr/local/lib/R/library"' is not writable
Would you like to use a personal library instead?  (y/n) y
Would you like to create a personal library
~/R/x86_64-pc-linux-gnu-library/3.2
to install packages into?  (y/n) 
```
It's unclear why Rstudio web interface is prompting for usage of `~/R/x86_64-pc-linux-gnu-library/3.2`.  Since, in it's session it has the user-space library.  See below screenshot : 

![screenshot from 2016-01-15 16 52 03](https://cloud.githubusercontent.com/assets/66993/12366319/686de710-bba8-11e5-9b4f-6ec189b8c0cc.png)

I'm going to leave this for now, and ask for comment.  Perhaps we should create a new directory in an EC2 instance, set it as the only library path for the ubuntu user / Rstudio, and then cut a new AMI from this further modification.  Any feedback is welcome.

cc/ @mtmorgan @vobencha @dtenenba

**Comment by b-long** [comment]
@priscian Helped me work through solving the problems described above.  We did change the ownership of `~/R-libs` , using the command `~/R-libs$ sudo chown -R ubuntu:ubuntu .`.

We also modified the `~/.Rprofile`, so that : 
1. `devtools` is loaded
2. Wrapped `biocLite()` to point to the `ubuntu` user's library (`~/R-libs`)
3. Replaced `.libPaths(c("~/R-libs", .libPaths()))` with `.libPaths("~/R-libs")`

We tested the environment (from a command-line R session and from the Rstudio web interface) and were able to successfully install packages and load packages.  After this, we created a new AMI.  It can be found here: [ami-56e2c53c](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-56e2c53c;sort=desc:creationDate)

**Comment by b-long** [comment]
An even newer AMI has been created (with feedback from Martin): [ami-2df1d647](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-2df1d647;sort=desc:creationDate)

**Comment by b-long** [comment]
The previous AMI, [ami-2df1d647](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-2df1d647;sort=desc:creationDate), (a.k.a. "_Bioconductor 3.2 Modified v2_") contained an error upon opening the Rstudio web interface.  This has been resolved by running `sudo service rstudio-server restart` before creating the AMI.  Here's the latest : [ami-56ab8c3c](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:search=ami-56ab8c3c).  This process should be added to the [standard procedures for creation of a Bioconductor AMI](https://github.com/Bioconductor/Bioconductor/wiki/AMI-management).

Unfortunately, due to a catastrophic failure on my home laptop, I don't have the ssh keypair needed in order to login to habu.  I'll need to test the rails application tomorrow.

**Comment by b-long** [comment]
It turns out Dan and I were wrong about avoiding the error message earlier (repeated below, for reference).  Invoking `sudo service rstudio-server restart` before cutting the AMI does not avoid this error message upon starting the rstudio web interface  : 

> The previous R session was abnormally terminated due to an unexpected crash.
> You may have lost workspace data as a result of this crash.

As such, we needed to create another AMI : "Bioconductor 3.2 Modified v4" , a.k.a. [_ami-7442641e_](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-7442641e;sort=desc:creationDate).  Luckily, there were further changes to [BiocIntroRPCI](https://github.com/Bioconductor/BiocIntroRPCI) that we needed to pull in as well.

The solution that does appear to work is the following : 
```
# Stop the service
service rstudio-server stop
# Since this doesn't stop all processes, need to find the remaining rstudio PID
ubuntu@ip-172-30-0-92:~$ sudo ps -ef |grep -i rstudio
sudo: unable to resolve host ip-172-30-0-92
ubuntu    1500     1  1 15:08 ?        00:00:00 /usr/lib/rstudio-server/bin/rsession -u ubuntu
ubuntu    1531  1476  0 15:09 pts/0    00:00:00 grep --color=auto -i rstudio
# Kill the rstudio rsession
kill 1500
# Remove the ~/.rstudio directory and the ~/.Rhistory file 
rm -rf .rstudio/ .Rhistory
```
After this, I've shutdown the instance and created the AMI from a cold state.
```
sudo shutdown -h now
```

**Comment by b-long** [comment]
For anyone following this discussion, the final AMI for the current course is here: [ami-d62305bc](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-d62305bc;sort=desc:creationDate), a.k.a. "_Bioconductor 3.2 Modified v5_" .  Information about AMI management is documented here: [AMI management](https://github.com/Bioconductor/Bioconductor/blob/master/documentation/AMI-management.md).

**Comment by dtenenba** [comment]
@b-long I see that the test course has not been updated with this AMI. Maybe that doesn't matter since the max instances are already running. Will you create a new course record for the actual course now, using the AMI above?

**Comment by b-long** [comment]
Yes, I'm going to create a new test right now (with 2 instances, based on [ami-d62305bc](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-d62305bc;sort=desc:creationDate)) .  The latest test will be : 
```ruby
new_course = Course.new(title: "Test 2016-JAN-20", location: 
"Buffalo, NY", startdate: "2016-01-20", enddate: "2016-01-20", ami_id: 
"ami-d62305bc", instance_type: "m3.large", max_instances: 2, password:
"public", gmt_offset: "-5")
new_course.save()
exit
```
Obviously we're confident this will work.  After observing it, I'll launch 45 instances, with similar configuration : 
```ruby 
new_course = Course.new(title: "Introduction to R / Bioconductor", location: 
"Buffalo, NY", startdate: "2016-01-20", enddate: "2016-01-22", ami_id: 
"ami-d62305bc", instance_type: "m3.large", max_instances: 45, password:
"public", gmt_offset: "-5")
new_course.save()
exit
```

**Comment by dtenenba** [comment]
OK. Maybe your new test should have a different `title` than the old one.
Thanks,
Dan

**Comment by b-long** [comment]
Done.  Feel free to login to (title) "_Test #2 2016-JAN-20_" with my credentials.

**Comment by b-long** [comment]
@mtmorgan found one last bug in the course AMI.  Specifically, the [extdata](https://github.com/Bioconductor/BiocIntroRPCI/tree/master/inst/extdata) must be available at `/home/ubuntu`.  I've since made that change and cut another AMI: "_Bioconductor 3.2 Modified v6_", a.k.a. [ami-a42b0dce](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-a42b0dce;sort=desc:creationDate) .

I have now deleted the course saved as _"Introduction to R / Bioconductor"_ and recreated it using the following rails console commands : 
```ruby
# Remove old course
old_course = Course.last
# Verify that this is the correct object
old_course
# If so, continue
old_course.delete()
# Construct new course
new_course = Course.new(title: "Introduction to R / Bioconductor", location: 
"Buffalo, NY", startdate: "2016-01-20", enddate: "2016-01-22", ami_id: 
"ami-a42b0dce", instance_type: "m3.large", max_instances: 45, password:
"public", gmt_offset: "-5")
new_course.save()
exit
```
Since this data is only needed for the course, we can do the following : 

1. Use Bioconductor 3.2 Modified v5 , a.k.a. [ami-d62305bc](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-d62305bc;sort=desc:creationDate) as the new 3.2 ami (replacing what is displayed at www.bioconductor.org/help/bioconductor-cloud-ami/#ami_ids ).
2. Use "_Bioconductor 3.2 Modified v6_", a.k.a. [ami-a42b0dce](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#Images:visibility=owned-by-me;search=ami-a42b0dce;sort=desc:creationDate) for the course and delete it once the course is finished.

Everything in production is now operational at : https://courses.bioconductor.org/

**Comment by dtenenba** [comment]
@b-long Please make sure that all instances associated with `old_course` are terminated, because now they will not be automatically terminated (doing that depends on there being a `Course` record in the database).
Thanks.

**Comment by b-long** [comment]
I believe I've already done that, but I'll double check shortly
